### PR TITLE
Vendor classes rewrite

### PR DESCRIPTION
**Pull request by cableman**
Vendor classes rewrite to better support development and new features (metrics).

### CODE REVIEW

**Review by turegjorup — COMMENTED** [review]
I thing we need to rethink having both `AbstractBaseVendorService` & `VendorServiceInterface`. I know why we're doing it. But it seems wrong to me. Once we have the `VendorServiceInterface` and vendors are must inherit from it the `VendorServiceInterface` becomes redundant. We might as well have the interface functions as abstract functions in the `VendorServiceInterface`

I think we solve the problems we are facing simply by having the `VendorCoreService`

**Review by turegjorup — COMMENTED** [review]
Looks good. But maybe change scope to be "vendor refactor". Then do metrics in a new PR?

**Comment by cableman** [comment]
Then I think what i'am missing here is moving Exception classes and check-up on the locks across processes

**Review by turegjorup — CHANGES_REQUESTED** [review]
I like this. I think we are close to it being done. Two things:

**Change requested:** 
`host` and `port` should be configurable in .env in `App\Service\MetricsService`

**Comment/suggestion:**
I don't like having the `getVendorId()` and `getVendorName()` duplicated in all vendor services. I know traits can't have constants but the can reference them. Also consolidating functions in the `VendorCoreService` is a big improvement but we could limit duplication even further by adding this:

`config/services.yaml`:
```
_instanceof:
    App\Service\VendorService\VendorServiceInterface:
        tags: [app.vendor.service]
            calls:
                - setVendorCoreService: [ '@App\Service\VendorService\VendorCoreService' ]
```

`App\Service\VendorService\VendorServiceInterface`:
```
    /**
     * Set the Vendor Core Service
     *
     * @param VendorCoreService $vendorCoreService
     *   The Vendor core service
     */
    public function setVendorCoreService(VendorCoreService $vendorCoreService): void;
```

`App\Service\VendorService\VendorServiceTrait`:
```
    private $vendorCoreService;

    /**
     * {@inheritdoc}
     */
    public function setVendorCoreService(VendorCoreService $vendorCoreService)
    {
        $this->vendorCoreService = $vendorCoreService;
    }

    /**
     * {@inheritdoc}
     */
    public function getVendorId(): int
    {
        return self::VENDOR_ID;
    }

    /**
     * {@inheritdoc}
     */
    public function getVendorName(): string
    {
        return $this->vendorCoreService->getVendorName($this->getVendorId());
    }
```

Then removing the functions from the individual vendor services.

File: src/Service/MetricsService.php [hunk 1]
```diff
@@ -24,7 +24,10 @@ class MetricsService
      */
     public function __construct()
     {
-        $adapter = new APC();
+        $adapter = new Redis([
+            'host' => 'redis',
+            'port' => 6379,
```
[the remark this answers is not in the document]
Comment (cableman) [reply src/Service/MetricsService.php:29]: dohh

[the remark this answers is not in the document]
Comment (cableman) [reply src/Service/MetricsService.php:29]: Focus was moved from that class after the talk about doing metrics in another PR :-) will fix

**Review by turegjorup — APPROVED** [review]

**Pull request merged**